README.md
# Node Pixel Async

> A completely derivative work of [node-pixel](https://github.com/ajfisher/node-pixel)

## Intent

I am using this repo as a place to make rapid changes, as it matures and I am able to add more meaningful updates to it I will update this readme. Until then consider this repo as unuseable unless you understand all of the code in it. I will be keeping it up to date with [node-pixel](https://github.com/ajfisher/node-pixel), and providing what patches I can. I'm trying to get something done by end of this month and iteration spped is vital!

Back to the README!

## Overview

The purpose of this library is to provide a node js interface for addressable
RGB LEDs. Most commonly these are known as Neo Pixels (if you shop at Adafruit)
however any WS2812b addressable LED should work with this library.

The current iteration supports two methods of set up:

* a custom version of firmata that provides an interface to talk to the "pixels".
* an I2C "backpack" using an arduino pro mini or nano that provides the interface
and control of the pixels and then the IO controller talks to this backpack over
I2C messages.

The pixel library can be used with both Johnny-Five or stock Node Firmata and
can be used by any board that provides an IO interface with I2C support such as
a Raspberry PI.

Both fimwares are provided in this repo in the
[firmware/build](firmware/build) directory.

## Getting help

If you need some help getting your pixel strip working with johnny five jump into
the [Gitter Chat](https://gitter.im/ajfisher/node-pixel) or reach out to
[ajfisher on twitter](http://twitter.com/ajfisher) or just raise an
[issue here.](https://github.com/ajfisher/node-pixel/issues)

## Installation

Installation of both backpack and custom firmata are covered in detail in the
[Installation Guide](docs/installation.md).

Short version for node-pixel custom firmata.

```
npm install node-pixel
npm install -g nodebots-interchange
```

Plug in your arduino

```
interchange install git+https://github.com/ajfisher/node-pixel -a uno --firmata
```

Note that on windows, you may need to explictly pass a port to flash due to the
way com ports work. eg:

```
interchange install git+https://github.com/ajfisher/node-pixel -a uno -p COM3 --firmata
```

### A note on multiple strips

Multiple led strips on one arduino or backpack are supported up to a maximum
of 8 individual strips (8 pins in use at once). Each strip can be different
lengths but you can only have a maximum of 192 pixels for Firmata and about
500 pixels for the backpack version.

Multiple strips connected to a single board or backpack are for the purposes
of node-pixel considered to be a single strip and are joined together in sequence
in the order that you define them.

On a backpack, the strips are defined sequentially from pin 0-7 on the backpack.

On an arduino, each strip can be defined with an individual pin which doesn't
need to be sequential (eg you can use pin 3, then pin 9, then pin 7).

One thing to note is that the timings on these strips are quite tight and you
will reach an upper limit of how much data you can push to the board controlling
the pixels (all that RGB data going over the wire) and the sheer number of
pixels you can refresh quickly (each pixel is written "in turn"). As such,
you may run into some blocking conflicts. These are discussed in
[this issue](https://github.com/ajfisher/node-pixel/issues/15).

# Pixel API

The Pixel API is provided below.

## Strip

A sequence of LEDs collected together is called a `strip`. A strip has
a controller to tell it to use the custom firmata or I2C backpack. A `strip`
can be a single physical strip in which case a single `pin` and `length` can
be provided. Otherwise it is made up of multiple physical strips, each of which
have their own `pin` and `length` and are composed together into order by using
the `strips` array as part of the definition of the object.

For the purposes of interaction however, once a `strip` is defined, it is all
one logical unit and the firmware will take care of writing data in the right
order, performing optimisations for strips that have or haven't changed and
writing in sequence or parallel as appropriate.

### Parameters

* **options** An object of property parameters

| Property | Type | Value / Description | Default | Required |
|----------|------|---------------------|---------|----------|
| pin | Number | Digital Pin. Used to set which pin the signal line to the pixels is being used when using a single strip. | 6 | no (4) |
| length | Number | Number of pixels to be set on a single strip or all strips if individual lengths are not defined in the `strips` array | 32 | no (5) |
| color_order | Constant | Determines the order of the RGB values against the pixels. Can be GRB, RGB or BRG  | pixel.COLOR_ORDER.GRB | no (6) |
| strips | Array | Array of pin and length objects or array of length objects | 6 | no (2)(3) |
| board | IO Object | IO Board object to use with Johnny Five | undefined | yes(1) |
| firmata | Firmata board | Firmata board object to use with Firmata directly | undefined | yes(1) |
| controller | String | I2CBACKPACK, FIRMATA | FIRMATA | no |
| skip_firmware_check | Boolean | If the controller is FIRMATA, optionally skip the check for the matching node-pixel sketch | false | no |
| gamma | Number | A number representing the gamma correction for a strip. Can be any decimal number. 2.8 generally works well. | 1.0 (7) | no |

(1) A board or firmata object is required but only one of these needs to be set.

(2) If using a backpack use an array of lengths eg `[8, 8, 8]` which would set
pins 0, 1 & 2 on the backpack to have strips of length 8 each on them.

(3) If using custom firmata then use an array of objects eg
`[ {pin: 4, length: 8}, {pin: 10, length: 8}, {pin: 11, length: 8} ]`
which would set pins 4, 10 & 11 to have strips of length 8 on each of them.

(4) If not supplied, it is assumed a `strips` array will be provided with a
`pin` parameter for each object in the array.

(5) If not supplied, it is assumed a `strips` array will be provided with a
`length` parameter for each object in the array.

(6) If supplied it will apply to all `strips` unless overridden selectively in
the `strips` array eg `[ {color_order: pixel.COLOR_ORDER.RGB}, ..]`

(7) Currently set to `1.0` to maintain current behaviour. Will move to `2.8` default
in version 0.10.

### Properties

* `length` - number of pixels in the `strip`
* `gamma` - the currently set gamma for the `strip`

### Events

* `onready()` -  emits when the `strip` is configured.
* `onerror()` - returns the error that occurred.

### Examples

_Johnny-Five instantiation_

```javascript
pixel = require("node-pixel");
five = require("johnny-five");

var board = new five.Board(opts);
var strip = null;

board.on("ready", function() {

    strip = pixel.Strip({
        board: this,
        controller: "FIRMATA",
        strips: [ {pin: 6, length: 4}, ], // this is preferred form for definition
        gamma: 2.8, // set to a gamma that works nicely for WS2812
    });

    strip.on("ready", function() {
        // do stuff with the strip here.
    });
});
```

_Firmata instantiation_

```javascript
pixel = require("node-pixel");
var firmata = require('firmata');

var board = new firmata.Board('path to usb',function(){

    strip = pixel.Strip({
        pin: 6, // this is still supported as a shorthand
        length: 4,
        firmata: board,
        controller: "FIRMATA",

    });

    strip.on("ready", function() {
        // do stuff with the strip here.
    });
});
```

_Johnny Five with backpack_

```javascript
pixel = require("node-pixel");
five = require("johnny-five");

var board = new five.Board(opts);

board.on("ready", function() {
    strip = pixel.Strip({
        board: this,
        controller: "I2CBACKPACK",
        strips: [4, 6, 8], // 3 physical strips on pins 0, 1 & 2 with lengths 4, 6 & 8.
    });

    strip.on("ready", function() {
        // do stuff with the strip here.

    });
});
```

Note that Johnny-Five uses the board option and firmata uses the firmata option.
This is because the pixel library supports a Board capable of presenting an
IO interface. The library will work out the right thing to do based on the
board being passed and the controller being set.

### Methods

#### show();

The show method should be called at the point you want to "set" the frame on
the strip of pixels and show them.

Note that when this method is called it will trigger the process that writes
the frame to the strips. If you have a very long strip of LEDs this may take
some time (assume 0.5ms per pixel) and is a blocking operation in most cases.

This gives you an upper limit as to how many frames you can drive per second.

Example

```javascript
// ... make pixel modifications

strip.show(); // make the strip latch and update the LEDs
```

Addressable LEDs work by clocking data along their entire length and so
you make the various changes you want to the strip as you need to without
triggering the display (like a frame buffer). Once you're ready you can then
call `show()` to propagate this data through the LEDs and display the frame.

#### off();

All LEDs on the strip can be turned off by using the `.off()` method. This effectively clears the current colours set on the strip.

`.clear()` is also aliased to the same method.

Example

```javascript
strip.off(); // turn the strip off / clear pixel colours
```

#### color( *colourstring* );

All LEDs on the strip can be set to the same colour using the `.color()` method.

`.colour()` is also aliased to the same method.

Parameters

* **colourstring** A `String` as a standard HTML hex colour or a CSS colour name,
or a CSS rgb(r, g, b) value used to specify the colour of the strip. Alternatively
an `Array` object as an rgb value eg `[r, g, b]`

Examples

_Set strip using a hex value_

```javascript
strip.color("#ff0000"); // turns entire strip red using a hex colour
strip.show();
```

_Update strip using a named CSS colour_

```javascript
strip.colour("teal"); // sets strip to a blue-green color using a named colour
strip.show();
```

_You can also use CSS RGB values_

```javascript
strip.color("rgb(0, 255, 0)"); // sets strip to green using rgb values
strip.show();
```

_Or set using an array of RGB values_

```javascript
strip.color([255, 255, 0]); // Sets strip using an array
strip.show();
```

#### shift( *amt*, *direction*, *wrap* );

All LEDs on the strip can be shifted along the strip forwards or backwards
by the given amount. This is very useful for long strip animation when you're
moving the whole strip by a pixel in one direction each frame and means you
don't have to send an update of `framelength` messages

Parameters

* **amt** A `Number` representing the number of pixels you want everything to
shift by.
* **direction** Either `Pixel.FORWARD` or `Pixel.BACKWARD` value which determines
direction of travel. Forward direction is in the flow index values (ie index 1->2 etc).
* **wrap** a `Boolean` representing whether to wrap the values that go off the
"end" of the strip back around to the start - useful for circular displays.

Example

```javascript
strip.pixel(0).color("#000");
strip.pixel(1).color("red");
strip.shift(1, pixel.FORWARD, true);
strip.pixel(1).color; // will now be nothing
strip.pixel(2).color; // will now be red.
```

#### pixel( *address* );

Individual pixels can be addressed by the pixel method using their address in
the sequence.

Note that if you have two physical strips of 8 and 10 then `pixel(10)` will be
the third pixel on the second physical strip.

Parameters

* **address** A `Number` indexing the pixel you want. Returns a `Pixel` object.

Example

```javascript
var p = strip.pixel(1); // get the second LED. p is now a Pixel object
```

## Pixel

A pixel is an individual element in the strip. It is fairly basic and it's API
is detailed below.

### Methods

#### color( *color string* )

Colors work exactly the same way on individual pixels as per strips so see the
`strip.color` reference above.

`.colour()` is aliased to this method as well.

Parameters

* **color string** A `String` providing the hex colour, CSS colour name or CSS
rgb() values to be used to set the individual pixel a certain colour. You can also
pass in an `Array` object that is a set of RGB values as [r, g, b].

Examples

```javascript
var p = strip.pixel(1);     // get second LED
p.color("#0000FF");         // set second pixel blue.

p = strip.pixel(2);         // get third LED
p.colour("orange");          // set third pixel red/yellow

p = strip.pixel(3);         // get fourth LED
p.color("rgb(0, 255, 0)");  // set fourth LED green

p = strip.pixel(4);         // get fifth LED
p.color([255, 0, 255]);     // set fifth LED magenta
```

#### color()

Returns an object representing the color of this pixel with the shape below.

`.colour()` is aliased to this method as well.

Parameters

* none

Shape

```javascript
{
    r: 0,               // red component
    g: 0,               // green component
    b: 0,               // blue component
    hexcode: "#000000", // hexcode of color
    color: "black",     // keyword name of color if matching
    rgb: [0,0,0],       // RGB component array
}
```

Example

Get a pixel, set it's colour and then query it's current state.

```javascript
var p = strip.pixel(1); // get second LED

p.color("#0000FF"); // set second pixel blue.

p.color(); // returns {r:0, g:0, b:255, hexcode:"#0000ff", color:"blue", rgb[0,0,255]}
```

#### off()

Turns the pixel to it's off state.

`.clear()` is also aliased to this method.

Example

Set a pixel value to off

```javascript

var p = strip.pixel(1); // get second LED
p.off(); // turn it off
p.color(); // returns {r: 0, g: 0, b: 0, hexcode:"#000000", color:"black", rgb: [0,0,0]}
strip.show(); // pixel will be off
```

## Detailed examples with circuits

* [Basic single strip using custom firmata](docs/firmata.md)
* [Basic single strip using Johnny Five Firmata](docs/johnnyfive.md)
* [Single strip using Johnny Five over I2C Backpack](docs/johnnyfive-i2c.md)
* [Mutiple strips using Johnny Five over firmata](docs/multipin.md)
* [Multiple strips using Johnny Five over I2C Backpack](docs/multipin-i2c.md)
* [Controlling a panel and using `strip.color()`](docs/panel.md)
* [Static rainbow with a single strip](docs/rainbow-static.md)
* [Dynamic rainbow over a single strip](docs/rainbow-dynamic.md)
* [Dynamic rainbow over multiple strips](docs/rainbow-dynamic-multipin.md)
* [Random pixels flowing down multiple strips](examples/mega-multipin.js)

## TODO and roadmap

This library is under active development and planned modifications are:

* Provide methods of having different shapes to the strips including 3D
* Prvide method of pixel selection using polar coordinates for circles and hexes


CHANGELOG.md
# [1.2.0](https://github.com/hweeks/node-pixel-async/compare/v1.1.0...v1.2.0) (2020-11-26)


### Bug Fixes

* allow color conversion based on channel maximums ([4f1d542](https://github.com/hweeks/node-pixel-async/commit/4f1d5425fd172063e07d80219fb8e5455102e1f7))


### Features

* **color:** allow color conversion based on channel maximums ([2bd4264](https://github.com/hweeks/node-pixel-async/commit/2bd42649e546a8953ec2e82eb036ea48e37de1e7))
* add base config for function ([ac5bac9](https://github.com/hweeks/node-pixel-async/commit/ac5bac9c1fbc9a6e73e924c51100f004fd34d2d4))

# [1.1.0](https://github.com/hweeks/node-pixel-async/compare/v1.0.0...v1.1.0) (2020-11-22)


### Features

* pixel breakdown to match strip ([137a623](https://github.com/hweeks/node-pixel-async/commit/137a623a921564008620cb4058da4ba07e914905))

# [1.0.0](https://github.com/hweeks/node-pixel-async/compare/v0.13.0...v1.0.0) (2020-11-22)


### Features

* **core:** move to ts completly, strip factory ([fc6e016](https://github.com/hweeks/node-pixel-async/commit/fc6e016f3719517d14560bf414db8ed11c241c1f))


### BREAKING CHANGES

* **core:** strip class becomes a factory, private api changes

# [0.13.0](https://github.com/hweeks/node-pixel-async/compare/v0.12.0...v0.13.0) (2020-11-21)


### Features

* typescript, linting ([510644e](https://github.com/hweeks/node-pixel-async/commit/510644e28420b88bbc30e0fa26d107da155aca93))
* update release process ([eb21cf1](https://github.com/hweeks/node-pixel-async/commit/eb21cf1c9f962b9f828eccf81afbfed3f1fdf2be))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.11.0](https://github.com/ajfisher/node-pixel/compare/v0.10.5...v0.11.0) (2020-10-25)


### Features

* **build:** Updated build process to use arduino-cli ([#175](https://github.com/ajfisher/node-pixel/issues/175)) ([effa990](https://github.com/ajfisher/node-pixel/commit/effa99034d6770419788eade327027d892dcd4dc))


### Build System

* **deps:** bump color-string from 1.5.3 to 1.5.4 ([3588aff](https://github.com/ajfisher/node-pixel/commit/3588affdc0679f1d927f44835fbe191c02d1b190))
* **deps-dev:** bump eslint from 7.10.0 to 7.11.0 ([1fd6af1](https://github.com/ajfisher/node-pixel/commit/1fd6af1283d131fcf988e2ffe41f91e7b31fa449))
* **deps-dev:** bump sinon from 9.1.0 to 9.2.0 ([#179](https://github.com/ajfisher/node-pixel/issues/179)) ([a9ea21f](https://github.com/ajfisher/node-pixel/commit/a9ea21f67988663e976fec931b3ee55e0ad5b0d5))
* Added github action for testing board compilation ([#176](https://github.com/ajfisher/node-pixel/issues/176)) ([5847799](https://github.com/ajfisher/node-pixel/commit/58477999099124921c50cfeedef9983e75ed1386))
* Move testing and coveralls to github actions ([#178](https://github.com/ajfisher/node-pixel/issues/178)) ([a49db6f](https://github.com/ajfisher/node-pixel/commit/a49db6faed2b59058a6be932c9c982b0f361b9c3)), closes [#177](https://github.com/ajfisher/node-pixel/issues/177)


### Documentation

* Updated readme to provide extra help for windows users. Fixes [#182](https://github.com/ajfisher/node-pixel/issues/182) ([#183](https://github.com/ajfisher/node-pixel/issues/183)) ([5e22e8a](https://github.com/ajfisher/node-pixel/commit/5e22e8af125582964c86cf7a8c1d8d9a377e03a7))


### Chores

* bump eslint from 7.11.0 to 7.12.0 ([#185](https://github.com/ajfisher/node-pixel/issues/185)) ([2dda10f](https://github.com/ajfisher/node-pixel/commit/2dda10f582ad2644dbe1906bb777e69b8c48bbc0))
* Create Dependabot config file ([#184](https://github.com/ajfisher/node-pixel/issues/184)) ([e390803](https://github.com/ajfisher/node-pixel/commit/e3908039898f48ae13fb53176116e2059b78fa82))

### [0.10.5](https://github.com/ajfisher/node-pixel/compare/v0.10.4...v0.10.5) (2020-10-05)


### Bug Fixes

* **linting:** Added linting checks to the JS files to the repo ([#169](https://github.com/ajfisher/node-pixel/issues/169)) ([f32c302](https://github.com/ajfisher/node-pixel/commit/f32c302517bb0581585dcf9f9b733462905d0fda))
* **utils:** Removed debug msg and added release recipe ([#172](https://github.com/ajfisher/node-pixel/issues/172)) ([710f29d](https://github.com/ajfisher/node-pixel/commit/710f29dea671ad24a5e4d28c71f1c84a2777f9e9))


### Chores

* **ci:** Add husky to repo with tests hooked correctly ([#170](https://github.com/ajfisher/node-pixel/issues/170)) ([e5e1a0e](https://github.com/ajfisher/node-pixel/commit/e5e1a0e500c9b878a07d456a194ce179ec2ce140))
* **ci:** Updated node versions to 12 and 14 for current support for travis ([#171](https://github.com/ajfisher/node-pixel/issues/171)) ([01f120e](https://github.com/ajfisher/node-pixel/commit/01f120ed665246a3af0d2f301cdf74d607c98ca4))


### Build System

* Additional make recipes to make life a bit easier ([#174](https://github.com/ajfisher/node-pixel/issues/174)) ([690d972](https://github.com/ajfisher/node-pixel/commit/690d9724341e999817642fc421228182bc404575))
* **docs:** Added changelog controls to standard-version to include other detail ([#173](https://github.com/ajfisher/node-pixel/issues/173)) ([0444aaa](https://github.com/ajfisher/node-pixel/commit/0444aaa1bc1d355907e1bb1d0bb7905a0bd57bb6))

### [0.10.4](https://github.com/ajfisher/node-pixel/compare/v0.10.3...v0.10.4) (2020-10-04)


### Bug Fixes

* **security:** Bump websocket-extensions from 0.1.3 to 0.1.4 ([#166](https://github.com/ajfisher/node-pixel/issues/166)) ([be57708](https://github.com/ajfisher/node-pixel/commit/be57708af67223001cbcc2cb997b1049a186b2f2))

### [0.10.3](https://github.com/ajfisher/node-pixel/compare/v0.10.2...v0.10.3) (2020-10-04)

### [0.10.2](https://github.com/ajfisher/node-pixel/compare/v0.10.0...v0.10.2) (2019-06-05)


### Bug Fixes

* **package:** update johnny-five to version 1.0.0 ([a43c8dc](https://github.com/ajfisher/node-pixel/commit/a43c8dc))



## 0.10.1

* Security updates to fix dependency security issues

## 0.10.0


lib/strip.js
const events = require('events');
const util = require('util');
const ColorString = require('color-string');
const { colorValue } = require('./pixel');
const {
  SHIFT_FORWARD
} = require('./constants');
const { Firmata } = require('./controllers/firmata');
const { IC2Backpack } = require('./controllers/backpack');

const strips = new WeakMap();

const Controllers = {
  FIRMATA: Firmata,
  I2CBACKPACK: IC2Backpack
};

function Strip(opts) {
  // opts contains an object with.
  // data: data pin for the pixel strip // DEPRECATED will be phased out.
  // length: length of the pixel strip. // DEPRECATED, will be phased out.
  // board: johnny five board object.
  // controller: controller type to use
  // firmata: actual firmata object if using firmata
  // stripShape: an array that contains lengths or optionally data pins and
  // lengths for each of them.
  //      eg: [ [6, 30], [12, 20], [7, 10] ] which would be 3 strips attached
  //      to pins 6, 12 and 7 and make a strip 60 pixels long.
  //      Otherwise [ 30, 20, 10 ] which would be 3 strips on PORTD 0-2 but
  //      still a strip 60 pixels long
  // gamma: A user specified value for gamma correction for the strip.
  //      default is 1.0 but will be changed to 2.8 over versions

  if (!(this instanceof Strip)) {
    return new Strip(opts);
  }

  let controller;

  if (typeof opts.controller === 'string') {
    controller = Controllers[opts.controller];
  } else {
    controller = opts.controller || Controllers['FIRMATA'];
  }

  this.dep_warning = {
    stripLength: false,
    gammaValue: (! typeof opts.gamma === 'undefined')
  };

  Object.defineProperties(this, controller);

  Object.defineProperty(this, 'length', {
    get() {
      const strip = strips.get(this);
      return strip.pixels.length;
    }
  });

  Object.defineProperty(this, 'gamma', {
    get() {
      const strip = strips.get(this);
      return strip.gamma;
    }
  });

  Object.defineProperty(this, 'gtable', {
    get() {
      const strip = strips.get(this);
      return strip.gtable;
    }
  });

  if (typeof this.initialize === 'function') {
    this.initialize(opts, strips);
  }
}

util.inherits(Strip, events.EventEmitter);

Strip.prototype.pixel = function(addr) {
  const strip = strips.get(this);

  return strip.pixels[addr];
};

Strip.prototype.colour = Strip.prototype.color = function(color, opts) {
  // sets the color of the entire strip
  // use a particular form to set the color either
  // color = hex value or named colors
  // or set color null and set opt which is an object as {rgb: [rx, gx, bx]}
  // values where x is an 8-bit value (0-255);
  const strip = strips.get(this);

  let stripcolor = null;

  if (color) {
    // use text to determine the color
    if (typeof(color) === 'object') {
      // we have an RGB array value
      stripcolor = color;
    } else {
      try {
        stripcolor = ColorString.get(color).value;
      } catch (e) {
        if (e instanceof TypeError && ColorString.get(color) === null ) {
          stripcolor = null;
        }
      }
    }
  }

  if (stripcolor != null) {
    // fill out the values for the pixels and then update the strip

    for (let i = 0; i < strip.pixels.length; i++) {
      strip.pixels[i].color(color, {sendmsg: false});
    }

    // set the whole strip color to the appropriate int value
    this.strip_color(colorValue(stripcolor, strip.gtable));
  } else {
    console.log("Supplied colour couldn't be parsed: " + color);
  }
}

Strip.prototype.off = Strip.prototype.clear = function() {
  // sets the strip to 'black', effectively setting it to 'off'
  this.color([0, 0, 0]);
  this.show();
};

Strip.prototype.shift = function(amt, direction, wrap) {
  // public version of the shift function independent of the controller.
  // this looks after the actual internal shifting of the pixels within the
  // js side and then calls the controller to mirror the same function.

  if (amt > 0) {
    const strip = strips.get(this);

    // take a copy of the pixels at the end that is being towards
    let start_element = 0;
    if (direction == SHIFT_FORWARD) {
      start_element = this.length - amt;
    }
    const tmp_pixels = strip.pixels.splice(start_element, amt);

    while (tmp_pixels.length > 0) {
      const px = tmp_pixels.pop();

      // set the pixel off if not wrapping.
      if (! wrap) {
        px.color('#000');
      }

      if (direction == SHIFT_FORWARD) {
        strip.pixels.unshift(px);
      } else {
        strip.pixels.push(px);
      }
    }

    // renumber the items so the addresses are correct for display
    strip.pixels.forEach((px, index) => {
      px.address = index;
    });

    // now get the firmware to update appropriately as well.
    this._shift(amt, direction, wrap);
  }
};

Strip.prototype.stripLength = function() {
  // gets the number of pixels in the strip
  if (! this.dep_warning.stripLength) {
    console.info('ERROR: strip.stripLength() is deprecated in favour of strip.length');
    console.info('0.8 - notice');
    console.info('0.9 - error');
    console.info('0.10 - removal');
    this.dep_warning.stripLength = true;
  }

  throw new Error({
    name: 'NotImplemented',
    message: 'stripLength is no longer supported, use strip.length',
    toString() { return 'NotImplemented: stripLength is no longer supported' }
  });
};

module.exports = {
  Strip
};
